apps/dotcom/sync-worker/src/UserDataSyncer.ts

```diff
@@ -0,0 +1,306 @@
+import {
+	OptimisticAppStore,
+	TlaRow,
+	ZEvent,
+	ZServerSentMessage,
+	ZStoreData,
+	ZTable,
+} from '@tldraw/dotcom-shared'
+import { assert, promiseWithResolve, sleep, uniqueId } from '@tldraw/utils'
+import { ExecutionQueue, createSentry } from '@tldraw/worker-shared'
+import postgres from 'postgres'
+import type { EventHint } from 'toucan-js/node_modules/@sentry/types'
+import { TLPostgresReplicator } from './TLPostgresReplicator'
+import {
+	fileKeys,
+	fileStateKeys,
+	getFetchUserDataSql,
+	parseResultRow,
+	userKeys,
+} from './getFetchEverythingSql'
+import { getPostgres } from './getPostgres'
+import { Environment } from './types'
+type PromiseWithResolve = ReturnType<typeof promiseWithResolve>
+
+export interface ZRowUpdateEvent {
+	type: 'row_update'
+	userId: string
+	// A unique id for the replicator's unbroken sequence of events.
+	// If the replicator fails and restarts, this id will change.
+	// And in that case the user DO should reboot.
+	sequenceId: string
+	row: TlaRow
+	table: ZTable
+	event: ZEvent
+}
+
+export interface ZBootCompleteEvent {
+	type: 'boot_complete'
+	userId: string
+	sequenceId: string
+	bootId: string
+}
+
+export interface ZMutationCommit {
+	type: 'mutation_commit'
+	userId: string
+	mutationNumber: number
+	// if the sequence id changes make sure we still handle the mutation commit
+	sequenceId: string
+}
+export interface ZForceReboot {
+	type: 'force_reboot'
+	sequenceId: string
+}
+
+export type ZReplicationEvent =
+	| ZRowUpdateEvent
+	| ZBootCompleteEvent
+	| ZMutationCommit
+	| ZForceReboot
+
+type BootState =
+	| {
+			type: 'init'
+			promise: PromiseWithResolve
+	  }
+	| {
+			type: 'connecting'
+			bootId: string
+			sequenceId: string
+			promise: PromiseWithResolve
+			bufferedEvents: Array<ZReplicationEvent>
+			didGetBootId: boolean
+			data: null | ZStoreData
+	  }
+	| {
+			type: 'connected'
+			bootId: string
+			sequenceId: string
+	  }
+
+export class UserDataSyncer {
+	state: BootState = {
+		type: 'init',
+		promise: promiseWithResolve(),
+	}
+	replicator: TLPostgresReplicator
+	db: postgres.Sql
+
+	store = new OptimisticAppStore()
+
+	sentry
+	// eslint-disable-next-line local/prefer-class-methods
+	private captureException = (exception: unknown, eventHint?: EventHint) => {
+		// eslint-disable-next-line @typescript-eslint/no-deprecated
+		this.sentry?.captureException(exception, eventHint) as any
+		if (!this.sentry) {
+			console.error(exception)
+		}
+	}
+
+	constructor(
+		private ctx: DurableObjectState,
+		private env: Environment,
+		private userId: string,
+		private broadcast: (message: ZServerSentMessage) => void
+	) {
+		this.sentry = createSentry(ctx, env)
+		this.replicator = env.TL_PG_REPLICATOR.get(
+			env.TL_PG_REPLICATOR.idFromName('0')
+		) as any as TLPostgresReplicator
+		this.db = getPostgres(env)
+		this.reboot(false)
```

Comment:
Not 100% on the naming here. `reboot` feels like something that happens after the boot already happened?